File: sites/sql/db/formatter.go
package db

import (
	"strings"
)

type MarkdownFormatter struct {
	Columns []string
	Rows    [][]string
}

func (f MarkdownFormatter) String() string {
	const verticalSeparator = `|`
	const horizontalSeparator = `-`

	// Find the longest datum per column to set an appropriate width for each column.
	var widths []int = make([]int, 0)
	for _, col := range f.Columns {
		widths = append(widths, len(col)+2) // Add two - we need at least one space on either side.
	}
	for _, row := range f.Rows {
		if len(f.Columns) != len(row) {
			return ""
		}
		for i, data := range row {
			widths[i] = max(widths[i], len(data)+2)
		}
	}
	var output strings.Builder

	// Populate the column names.
	for i, col := range f.Columns {
		remaining := widths[i] - len(col)
		space, extra := remaining/2, remaining%2
		_, _ = output.WriteString(verticalSeparator)
		_, _ = output.WriteString(strings.Repeat(` `, space))
		_, _ = output.WriteString(col)
		_, _ = output.WriteString(strings.Repeat(` `, space+extra))
	}
	_, _ = output.WriteString(verticalSeparator)
	_, _ = output.WriteString("\n")

	// Create a column header separator.
	for i := range f.Columns {
		_, _ = output.WriteString(verticalSeparator)
		for range widths[i] {
			_, _ = output.WriteString(horizontalSeparator)
		}
	}
	_, _ = output.WriteString(verticalSeparator)
	_, _ = output.WriteString("\n")

	// Add the rows.
	for _, row := range f.Rows {
		for i, data := range row {
			remaining := widths[i] - len(data) - 1 // Left justify the data.
			_, _ = output.WriteString(verticalSeparator)
			_, _ = output.WriteString(` `)
			_, _ = output.WriteString(data)
			_, _ = output.WriteString(strings.Repeat(` `, remaining))
		}
		_, _ = output.WriteString(verticalSeparator)
		_, _ = output.WriteString("\n")
	}

	return output.String()
}

type BoxFormatter struct {
	Columns []string
	Rows    [][]string
}

// TODO: Implement.
// https://en.wikipedia.org/wiki/Box-drawing_characters
func (f BoxFormatter) String() string {
	const verticalSeparator = `|`
	const horizontalSeparator = `-`
	if len(f.Columns) != len(f.Rows) {
		return ""
	}
	return ""
}

File: sites/sql/www/src/db/formatter.ts
import { QueryExecResult } from "sql.js";

export class MarkdownFormatter {
  columns: string[];
  rows: string[][];

  private constructor() {}

  static fromResult(result: QueryExecResult): MarkdownFormatter {
    var mf = new MarkdownFormatter();
    mf.columns = result.columns;
    mf.rows = result.values.map((row) =>
      row.map((value) => (value === null ? "NULL" : value.toString())),
    );
    return mf;
  }

  toString(): string {
    const verticalSeparator = `|`;
    const horizontalSeparator = `-`;

    // Find the longest datum per column to set an appropriate width for each column.
    var widths: number[];
    widths = this.columns.map((col) => col.length + 2); // Add two - we need at least one space on either side.
    this.rows.map((row) =>
      row.map((data, i) => (widths[i] = Math.max(widths[i], data.length + 2))),
    );

    let output = "";

    // Populate the column names.
    this.columns.map((col, i) => {
      let remaining = widths[i] - col.length;
      let space = remaining / 2;
      let extra = remaining % 2;
      output += verticalSeparator;
      output += " ".repeat(space);
      output += col;
      output += " ".repeat(space + extra);
      // if (i < this.columns.length - 1) {
      //   output += verticalSeparator;
      // }
    });
    output += verticalSeparator;
    output += "\n";

    // Create a column header separator.
    this.columns.map((col, i) => {
      output += verticalSeparator;
      for (let j = 0; j < widths[i]; j++) {
        output += horizontalSeparator;
      }
      // if (i < this.columns.length - 1) {
      //   output += verticalSeparator;
      // }
    });
    output += verticalSeparator;
    output += "\n";

    // Add the rows.
    for (let row of this.rows) {
      row.map((data, i) => {
        let remaining = widths[i] - data.length - 1; // Left justify the data.
        output += verticalSeparator;
        output += " ";
        output += data;
        output += " ".repeat(remaining);
        // if (i < row.length - 1) {
        //   output += verticalSeparator;
        // }
      });
      output += verticalSeparator;
      output += "\n";
    }

    return output;
  }
}
